cmd/kubectl-gadget/trace/tcpdump.go

```diff
@@ -0,0 +1,399 @@
+				decoderBinary = "wireshark"
+			case DecoderTCPDump:
+				decoder = DecoderExternal
+				decoderArgs = []string{"-r", "-"}
+				decoderBinary = "tcpdump"
+			case DecoderInternal:
+				decoder = DecoderInternal
+			case DecoderFile:
+				decoder = DecoderFile
+			default:
+				return errors.New("unknown decoder")
+			}
+
+			if decoder == DecoderFile {
+				if filenameParam == "" {
+					return fmt.Errorf("no filename specified")
+				}
+				f, err := os.Create(filenameParam)
+				if err != nil {
+					return fmt.Errorf("creating file: %w", err)
+				}
+				out = f
+			}
+
+			if decoder == DecoderExternal {
+				r, w, err := os.Pipe()
+				if err != nil {
+					return fmt.Errorf("could not create pipe: %w", err)
+				}
+
+				if decoderArgsParam != "" {
+					decoderArgs = append(decoderArgs, strings.Split(decoderArgsParam, " ")...)
+				}
+				if decoderBinaryParam != "" {
+					decoderBinary = decoderBinaryParam
+				}
+
+				// xwr := &wr{File: os.Stdout}
+
+				decoderCmd = exec.Command(decoderBinary, decoderArgs...)
+				decoderCmd.Stdout = os.Stdout
+				decoderCmd.Stderr = os.Stderr
+				decoderCmd.Stdin = r
+				err = decoderCmd.Start()
+				if err != nil {
+					return fmt.Errorf("could not start tcpdump: %w", err)
+				}
+
+				out = w
+			}
+
+			if out != nil {
+				var err error
+				dummyInterface := pcapgo.DefaultNgInterface
+				dummyInterface.LinkType = layers.LinkTypeEthernet
+				dummyInterface.SnapLength = uint32(snapLen)
+				ngw, err = pcapgo.NewNgWriterInterface(out, dummyInterface, pcapgo.NgWriterOptions{SectionInfo: pcapgo.NgSectionInfo{
+					Hardware:    runtime.GOARCH,
+					OS:          runtime.GOOS,
+					Application: "InspektorGadget",
+					Comment:     "using gopacket",
+				}})
+				if err != nil {
+					return fmt.Errorf("instantiating NgWriter: %w", err)
+				}
+				ngw.Flush()
+			}
+
+			filter := strings.Join(args, " ")
+
+			tcpdumpGadget := &TraceGadget[types.Event]{
+				name:        "tcpdump",
+				commonFlags: commonFlags,
+				parser:      NewTCPDump(&commonFlags.OutputConfig, filter, snapLen, decoder, ngw, out),
+				params: map[string]string{
+					types.FilterStringParam: filter,
+					types.SnapLenParam:      strconv.Itoa(snapLen),
+				},
+			}
+
+			return tcpdumpGadget.Run()
+		},
+	}
+
+	utils.AddCommonFlags(cmd, commonFlags)
+	cmd.Flags().StringVar(&decoderParam, "decoder", "internal", "name of the decoder to use (either tcpdump, wireshark, internal or file)")
+	cmd.Flags().StringVar(&decoderArgsParam, "decoder-args", "", "arguments to forward to decoder")
+	cmd.Flags().StringVar(&decoderBinaryParam, "decoder-binary", "", "path to decoder binary (defaults to 'wireshark' or 'tcpdump' depending on decoder)")
+	cmd.Flags().StringVar(&filenameParam, "out-file", "", "output file name")
+	cmd.Flags().BoolVar(&disableDNSPopulation, "disable-dns", false, "disable DNS population from kubernetes")
+	cmd.Flags().IntVar(&snapLen, "snaplen", 68, "number of bytes to capture per packet")
```

Comment:
While tracing DNS requests I am seeing error because of packet truncation:
```
PACKET: 68 bytes, truncated
- Layer 1 (14 bytes) = Ethernet	{Contents=[..14..] Payload=[..54..] SrcMAC=02:42:ac:11:00:03 DstMAC=02:42:ac:11:00:02 EthernetType=IPv4 Length=0}
- Layer 2 (20 bytes) = IPv4	{Contents=[..20..] Payload=[..34..] Version=4 IHL=5 TOS=0 Length=175 Id=34611 Flags=DF FragOffset=0 TTL=64 Protocol=UDP Checksum=64653 SrcIP=10.96.0.10 DstIP=172.17.0.2 Options=[] Padding=[]}
- Layer 3 (08 bytes) = UDP	{Contents=[..8..] Payload=[..26..] SrcPort=53(domain) DstPort=54502 Length=155 Checksum=46889}
- Layer 4 (26 bytes) = DecodeFailure	Packet decoding error: dns name uncomputable: invalid index
```
Setting snaplen to higher values gets rid of errors.

Reply:
I first thought these would be panics - but those errors are just coming from truncated packets, so I think this is fine.